### PR TITLE
[SemanticDB] Fix missing symbol occurrence of type params' type bounds for constructor

### PR DESCRIPTION
**Pull request by tanishiking**
Fix https://github.com/lampepfl/dotty/issues/13270

This PR fixes missing symbol occurrence of

- type parameters' upper bounds of constructor https://github.com/lampepfl/dotty/pull/13284/commits/5d3bdf3583ed04f401cd283f6f1c73d2e5ddff37
- TypeRef's symbol inside of TypeTree https://github.com/lampepfl/dotty/pull/13284/commits/06872697db9cf4b369be6835ee41b7c8dbacc028

---

For example:

```scala
// Missing symbol occurrence of `<: Txn[T]`
trait Txn/*<-_empty_::Txn#*/[T/*<-_empty_::Txn#[T]*/ <: Txn[T]]
```

The root problem is we are ignoring TypeDef's rhs of type parameters in the constructor. Once it traverses the trees (like `TypeBoundsTree` and `AppliedTypeTree`), the original traverser will look into `Ident(Txn)` and `Ident(T)` inside of `AppliedTypeTree` in the following case.

```scala
// AST of `trait Txn[T <: Txn[T]]`
TypeDef(
  Txn,
  Template(
    DefDef(
      name <init>,
      paramss = List(
        List(
          TypeDef(
            T,
            TypeBoundsTree(
              TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing)],
              AppliedTypeTree(Ident(Txn),List(Ident(T))),EmptyTree
            )
          )
        ),
        List()
      ),
   ...
)
```

---

Also registered symbol that appears inside of TypeRef (of TypeTree).
I understand TypeTree can contain any type, but as far as I know, `TypeRef` is the only type that is present in the source and we need to handle.

### CODE REVIEW

File: tests/semanticdb/expect/i9782.expect.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// LazyRef
+trait Txn/*<-_empty_::Txn#*/[T/*<-_empty_::Txn#[T]*/ <: Txn/*->_empty_::Txn#*/[T/*->_empty_::Txn#[T]*/]]
+
+trait Elem/*<-_empty_::Elem#*/[T/*<-_empty_::Elem#[T]*/ <: Txn/*->_empty_::Txn#*/[T/*->_empty_::Elem#[T]*/]]
+
+trait Obj/*<-_empty_::Obj#*/[T/*<-_empty_::Obj#[T]*/ <: Txn/*->_empty_::Txn#*/[T/*->_empty_::Obj#[T]*/]] extends Elem/*->_empty_::Elem#*/[T/*->_empty_::Obj#[T]*/]
+
+trait Copy/*<-_empty_::Copy#*/[In/*<-_empty_::Copy#[In]*/ <: Txn/*->_empty_::Txn#*/[In/*->_empty_::Copy#[In]*/], Out/*<-_empty_::Copy#[Out]*/ <: Txn/*->_empty_::Txn#*/[Out/*->_empty_::Copy#[Out]*/]] {
+  def copyImpl/*<-_empty_::Copy#copyImpl().*/[Repr/*<-_empty_::Copy#copyImpl().[Repr]*/[~ <: Txn[~]] <: Elem[~]](in/*<-_empty_::Copy#copyImpl().(in)*/: Repr/*->_empty_::Copy#copyImpl().[Repr]*/[In/*->_empty_::Copy#[In]*/]): Repr/*->_empty_::Copy#copyImpl().[Repr]*/[Out/*->_empty_::Copy#[Out]*/]
```
Comment (bishabosha) [inline tests/semanticdb/expect/i9782.expect.scala:9]: it looks like here that there are no occurrences for the parameters of `Repr`

**Review by bishabosha — DISMISSED** [review]

File: tests/semanticdb/expect/i9782.expect.scala (hunk 1 above)
Comment (bishabosha) [reply tests/semanticdb/expect/i9782.expect.scala:9]: although looking at the tree it seems maybe something else is hiding the occurrence:
```scala
-- Info: /Users/jamie/Workspace/dotty/tests/semanticdb/expect/i9782.scala:9:6 --
9 |  def copyImpl[Repr[~ <: Txn[~]] <: Elem[~]](in: Repr[In]): Repr[Out]
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |here is a defdef tree repr: DefDef(copyImpl,List(List(TypeDef(Repr,LambdaTypeTree(List(TypeDef(~,TypeBoundsTree(TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing)],AppliedTypeTree(Ident(Txn),List(Ident(~))),EmptyTree))),TypeBoundsTree(TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing)],AppliedTypeTree(Ident(Elem),List(Ident(~))),EmptyTree)))), List(ValDef(in,AppliedTypeTree(Ident(Repr),List(Ident(In))),EmptyTree))),AppliedTypeTree(Ident(Repr),List(Ident(Out))),EmptyTree)
```

Comment (bishabosha) [reply tests/semanticdb/expect/i9782.expect.scala:9]: this could be addressed separately

Comment (tanishiking) [reply tests/semanticdb/expect/i9782.expect.scala:9]: Oh, right. I think we can fix it in this PR, I'll take a look :)

**Comment by bishabosha** [comment]
This will still need a rebase

**Review by bishabosha — APPROVED** [review]
Thanks again!

File: tests/semanticdb/expect/i9782.expect.scala (hunk 1 above)
Comment (tanishiking) [reply tests/semanticdb/expect/i9782.expect.scala]: I found type params are going to this branch:
https://github.com/lampepfl/dotty/blob/28a671a71a3a5f6627a2db0c9cc5d9c0ec6218a2/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala#L168-L170

and `excludeChildren` is true if the parameter is HKTypeLambda, and that's why those symbols are missing.

**Comment by tanishiking** [comment]
Fixed failed rebase :bow: https://github.com/lampepfl/dotty/pull/13284/commits/28a671a71a3a5f6627a2db0c9cc5d9c0ec6218a2

**Review by bishabosha — APPROVED** [review]

**Pull request merged**